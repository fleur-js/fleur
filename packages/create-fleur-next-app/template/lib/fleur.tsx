import { default as NextApp } from 'next/app'
import { FleurContext } from '@fleur/react'
import { useMemo } from 'react'
import { bindFleurContext } from '@fleur/next'
import { createContext } from '../domains/app'

export const getOrCreateFleurContext = (state: any = null) => {
  const isServer = typeof window === 'undefined'
  const context = createContext()

  if (state) {
    context.rehydrate(state)
  }

  return context
}

export const appWithFleurContext = (App: typeof NextApp) => {
  const Comp = ({ __FLEUR_STATE__, ...props }: any) => {
    const fleurContext = useMemo(
      () => getOrCreateFleurContext(__FLEUR_STATE__),
      [],
    )

    return (
      <FleurContext value={fleurContext}>
        <App {...props} />
      </FleurContext>
    )
  }

  Comp.getInitialProps = async appContext => {
    const fleurContext = getOrCreateFleurContext()
    bindFleurContext(fleurContext, appContext)

    const appProps = await App.getInitialProps(appContext)

    return {
      ...appProps,
      __FLEUR_STATE__: fleurContext.dehydrate(),
    }
  }

  return Comp
}
