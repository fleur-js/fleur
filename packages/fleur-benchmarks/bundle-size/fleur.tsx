import Fleur, { action, operations, reducerStore } from '@fleur/fleur'
import { useFleurContext } from '@fleur/react'
import { FleurContext, useStore } from '@fleur/react'
import { useCallback } from 'react'
import { render } from 'react-dom'

const appActions = {
  fetchCount: action.async<void, { count: number }>(),
  increment: action<{ amount: number }>(),
  decrement: action<{ amount: number }>(),
}

const AppStore = reducerStore('App', () => ({ count: 0 }))
  .listen(
    appActions.fetchCount.done,
    (draft, { count }) => (draft.count = count),
  )
  .listen(appActions.increment, (draft, { amount }) => {
    draft.count += amount
  })
  .listen(appActions.decrement, (draft, { amount }) => {
    draft.count -= amount
  })

const AppOps = operations({
  fetchCount: async (context, { userId }: { userId: number }) => {
    context.dispatch(appActions.fetchCount.done, { count: 1 })
  },
  increment: (context, { amount }: { amount: number }) => {
    context.dispatch(appActions.increment, { amount })
  },
  decrement: (context, { amount }: { amount: number }) => {
    context.dispatch(appActions.decrement, { amount })
  },
})

const app = new Fleur({ stores: [AppStore] })
const context = app.createContext()

;(async () => {
  await context.executeOperation(AppOps.fetchCount, { userId: 1 })
  await context.executeOperation(AppOps.increment, { amount: 10 })

  console.log("It's Fleur")
  console.log('Assert', context.getStore(AppStore).state.count === 11)
})()

const App = () => {
  const { executeOperation } = useFleurContext()
  const count = useStore((get) => get(AppStore).state.count)
  const handleClick = useCallback(() => {
    executeOperation(AppOps.increment, { amount: 1 })
  }, [])

  return <button onClick={handleClick}>{count}</button>
}

document.addEventListener('DOMContentLoaded', () => {
  render(
    <FleurContext value={context}>
      <App />
    </FleurContext>,
    document.getElementById('root'),
  )
})
