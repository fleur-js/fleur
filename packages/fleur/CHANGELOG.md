### 0.0.7
### API Changes
- Accept action name in first argument of `action()`
  ``` typescript
  const increase = action<{ amount: number }>('increase')
  ```
- Add `actions()` for auto naming actions
  ```typescript
  export const CountActions = actions('Count', {
      increase: action<{ amount: number }>(), // It names 'Count/increase'
  })
  ```
- `ExtractActionIdentifiersFromObject` renamed to `ActionsOf`

### 0.0.6
#### API Changes
- Allow to any types in Store#rehydrate,dehydrate

### 0.0.5
#### Fixes
- Fix missing async-process waiting for OperationContext#e
xecuteOperation

### 0.0.4
#### API Changes
- Accept storeName: string in AppContext.getStore for improve debuggerbility
  No supports in another context getStore methods. (Usable only AppContext.getStore)

### 0.0.3
#### API Changes
- Accept many listener per Action in one Store
  ```ts
  // invalid in 0.0.2
  class SomeStore extends Store {
      private handleSomeAction = listen(someAction, () => { /* Do something */ })
      private handleSomeAction2 = listen(someAction, () => { /* Do something */ }) // error!
  }

  // Can in 0.0.3
  class SomeStore extends Store {
      private handleSomeAction = listen(someAction, () => { /* Do something */ })
      private handleSomeAction2 = listen(someAction, () => { /* Do something */ })
  }
  ```

### 0.0.2
#### API Changes
- Start changelog.
- Rename `Store#produce` to `Store#updateWith`.
- `static storeName = 'someStoreName'` is required for Store.

#### Fixes
- Broken store rehydration state under the Uglify
