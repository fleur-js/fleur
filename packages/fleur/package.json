{
  "name": "@ragg/fleur",
  "version": "0.0.8",
  "homepage": "https://github.com/ra-gg/fleur/tree/master/packages/fleur",
  "main": "lib/index.js",
  "types": "typings/index.d.ts",
  "author": "Ragg<ragg.devpr@gmail.com>",
  "license": "MIT",
  "scripts": {
    "prepublishOnly": "rm -rf ./{typings,lib} && tsc --emitDeclarationOnly --declaration --outDir ./typings && tsc --outDir lib",
    "test": "jest --verbose"
  },
  "devDependencies": {
    "@types/eventemitter3": "^2.0.2",
    "@types/flux-standard-action": "1.1.0",
    "@types/invariant": "2.2.29",
    "cache-loader": "1.2.2",
    "eventemitter3": "3.1.0",
    "husky": "1.1.2",
    "jest": "23.6.0",
    "lint-staged": "7.3.0",
    "ts-jest": "23.10.4",
    "typescript": "3.1.3"
  },
  "dependencies": {
    "immer": "1.7.2",
    "invariant": "2.2.4"
  }
}
