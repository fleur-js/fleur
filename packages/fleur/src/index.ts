import Fleur from './Fleur'

export { Fleur as default } // Aliasing for editor intellisense
export { default as Store, StoreClass, listen } from './Store'
export { default as OperationContext } from './OperationContext'
export { action, actions, ActionsOf } from './Action'
export { default as AppContext } from './AppContext'
export { default as ComponentContext } from './ComponentContext'
export { operations, operation } from './Operations'
