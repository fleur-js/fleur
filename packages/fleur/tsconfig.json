{
    "compilerOptions": {
        "baseUrl": "./src",
        "target": "es2015",
        "module": "commonjs",
        "moduleResolution": "node",
        "noUnusedLocals": true,
        "noUnusedParameters": true,
        "noImplicitAny": true,
        "noImplicitReturns": true,
        "strictNullChecks": true,
        "forceConsistentCasingInFileNames": true,
        "lib": [
            "es2015",
            "es2016",
            "es2017",
            "dom"
        ],
        "outDir": "dist",
    },
    "exclude": [
        "src/**/*.spec.ts",
        "test_lib/"
    ]
}
