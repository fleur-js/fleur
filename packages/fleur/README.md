# ðŸŒ¼ Fleur ðŸŒ¼ [![npm version](https://badge.fury.io/js/%40ragg%2Ffleur.svg)](https://www.npmjs.com/package/@ragg/fleur) [![travis](https://travis-ci.org/ra-gg/fleur.svg?branch=master)](https://travis-ci.org/ra-gg/fleur)
An Fully-typed Flux framework inspired by Fluxible.
Runs on Node / Web.

(No dependence to React. see [this](https://www.npmjs.com/package/@ragg/fleur-react) if you want to use with React.)

## Example

``` typescript
// actions.ts (Action typings)
import { action } from '@ragg/fleur'

export const CounterActions = actions('CounterAction', {
    increase: action<{ amount: number }>(),
    decrease: action<{ amount: number }>(),
}
```

``` typescript
// store.ts (Store)
import { listen, Store } from '@ragg/fleur'
import { CounterActions } from './actions.ts'

export default class SomeStore extends Store {
    public state: { count: number } = { count: 0 }

    private handleIncrease = listen(CounterActions.increase, (payload) => {
        // `this.updateWith` is immutable changing `this.state` with `immer.js`
        this.updateWith(draft => draft.count += payload.amount)
    })

    private handleDecrease = listen(CounterActions.decrease, (payload) => {
        this.updateWith(draft => draft.count -= payload.amount)
    })

    public getCount() {
        return this.state.count
    }
}
```

``` typescript
// operations.ts (Action Creator)
import { operation } from '@ragg/fleur'
import { CounterActions } from './actions.ts'

export const increaseOperation = operation((ctx, { amount }: { amount: number }) => {
    ctx.dispatch(CounterActions.increase, { amount })
})

export const decreaseOperation = operation((ctx, { amount }: { amount: number }) => {
    ctx.dispatch(CounterActions.decrease, { amount })
})
```

``` typescript
// app.ts
import Fleur from '@ragg/fleur'
import SomeStore from './store.ts'

const app = new Fleur({
    stores: [SomeStore],
})

const ctx = app.createContext()
ctx.executeOperation(increaseOperation, { amount: 10 })
console.log(ctx.getStore(SomeStore).getCount()) // => 10
```

## How to use with React?
See [`@ragg/fleur-react`](https://www.npmjs.com/package/@ragg/fleur-react).
