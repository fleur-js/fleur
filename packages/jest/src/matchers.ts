import { ActionIdentifier, Operation } from '@fleur/fleur'
import { MockedOperationContext, MockedComponentContext } from '@fleur/testing'
import * as deepEqual from 'fast-deep-equal'
import { matcherHint, printDiffOrStringify } from 'jest-matcher-utils'

export const toDispatchAction = (
  context: MockedOperationContext,
  expectedAction: ActionIdentifier<any>,
  expectedPayload: any,
): jest.CustomMatcherResult => {
  let actionMatchedPayload: any = null

  for (const dispatch of context.dispatches) {
    if (dispatch.action === expectedAction) {
      actionMatchedPayload = dispatch.payload

      if (deepEqual(dispatch.payload, expectedPayload)) {
        return {
          pass: true,
          message: () => `Expect to dispatch action \`${expectedAction.name}\``,
        }
      }

      continue
    }
  }

  return {
    pass: false,
    message: actionMatchedPayload
      ? () =>
          matcherHint(
            '.toDispatchAction',
            'context',
            `${expectedAction.name}, payload`,
          ) +
          `\nAction is dispatched, but payload is different.\n` +
          printDiffOrStringify(
            expectedPayload,
            actionMatchedPayload,
            'Expected',
            'Actual',
            true,
          )
      : () =>
          `Expect to dispatch action \`${expectedAction.name}\` but not dispatched`,
  }
}

export const toExecuteOperation = (
  context: MockedComponentContext,
  expectedOp: Operation,
  expectedArgs: any[],
): jest.CustomMatcherResult => {
  let actualArguments: any = null

  for (const execute of context.executes) {
    if (execute.op === expectedOp) {
      actualArguments = execute.args

      if (deepEqual(execute.args, expectedArgs)) {
        return {
          pass: true,
          message: () => `Expect to execute operation \`${execute.op.name}\``,
        }
      }

      continue
    }
  }

  return {
    pass: false,
    message: actualArguments
      ? () =>
          matcherHint('.toExecOperation', 'context', `op, ...args`) +
          `\nOperation is executed, but arguments is different.\n` +
          printDiffOrStringify(
            expectedArgs,
            actualArguments,
            'Expected',
            'Actual',
            true,
          )
      : () =>
          `Expect to execute operation \`${expectedOp.name}\` but not executed.`,
  }
}
