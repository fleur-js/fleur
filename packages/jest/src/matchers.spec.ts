import './index'
import { reducerStore, operation, action } from '@fleur/fleur'
import { mockFleurContext } from '@fleur/testing'
import { mockStore } from '@fleur/testing'
import { toDispatchAction, toExecuteOperation } from './matchers'

describe('matchers', () => {
  const Store = reducerStore('Store', () => ({}))
  const act = action<{ test: string }>('action')
  const op = operation((context, text: string) =>
    context.dispatch(act, { test: text }),
  )

  const baseContext = mockFleurContext({
    stores: [mockStore(Store, {})],
  })

  describe('toDispatchAction', () => {
    it('Passed', async () => {
      const context = baseContext.derive().mockOperationContext()
      await context.executeOperation(op, 'hi')

      const result = toDispatchAction(context, act, { test: 'hi' })
      expect(result.pass).toBe(true)
      expect(result.message()).toMatchInlineSnapshot(
        `"Expect to dispatch action \`action\`"`,
      )
    })

    it('Action matched but different payload', async () => {
      const context = baseContext.derive().mockOperationContext()
      await context.executeOperation(op, 'hi')

      const result = toDispatchAction(context, act, { test: 'ha?' })
      expect(result.pass).toBe(false)
      expect(result.message()).toMatchSnapshot()
    })

    it('Action not matched', async () => {
      const unmatchedAct = action<{}>('unmatchedAction')
      const context = baseContext.derive().mockOperationContext()
      await context.executeOperation(op, 'hi')

      const result = toDispatchAction(context, unmatchedAct, { test: 'ha?' })
      expect(result.pass).toBe(false)
      expect(result.message()).toMatchInlineSnapshot(
        `"Expect to dispatch action \`unmatchedAction\` but not dispatched"`,
      )
    })
  })

  describe('toExecOperation', () => {
    it('Passed', async () => {
      const context = baseContext.derive().mockComponentContext()
      await context.executeOperation(op, 'hi')

      const result = toExecuteOperation(context, op, ['hi'])
      expect(result.pass).toBe(true)
      expect(result.message()).toMatchInlineSnapshot(
        `"Expect to execute operation \`\`"`,
      )
    })

    it('Operation matched but different payload', async () => {
      const context = baseContext.derive().mockComponentContext()
      await context.executeOperation(op, 'hi')

      const result = toExecuteOperation(context, op, ['ha?'])
      expect(result.pass).toBe(false)
      expect(result.message()).toMatchSnapshot()
    })

    it('Operation not matched', async () => {
      const unmatchedOp = operation(() => {})
      const context = baseContext.derive().mockComponentContext()
      await context.executeOperation(op, 'hi')

      const result = toExecuteOperation(context, unmatchedOp, ['hi'])
      expect(result.pass).toBe(false)
      expect(result.message()).toMatchInlineSnapshot(
        `"Expect to execute operation \`\` but not executed."`,
      )
    })
  })
})
