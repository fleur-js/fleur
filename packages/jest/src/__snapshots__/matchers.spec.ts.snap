// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matchers toDispatchAction Action matched but different payload 1`] = `
"[2mexpect([22m[31mcontext[39m[2m).toDispatchAction([22m[32maction, payload[39m[2m)[22m
Action is dispatched, but payload is different.
[32m- Expected  - 1[39m
[31m+ Actual    + 1[39m

[2m  Object {[22m
[32m-   \\"test\\": \\"ha?\\",[39m
[31m+   \\"test\\": \\"hi\\",[39m
[2m  }[22m"
`;

exports[`matchers toExecOperation Operation matched but different payload 1`] = `
"[2mexpect([22m[31mcontext[39m[2m).toExecOperation([22m[32mop, ...args[39m[2m)[22m
Operation is executed, but arguments is different.
[32m- Expected  - 1[39m
[31m+ Actual    + 1[39m

[2m  Array [[22m
[32m-   \\"ha?\\",[39m
[31m+   \\"hi\\",[39m
[2m  ][22m"
`;
